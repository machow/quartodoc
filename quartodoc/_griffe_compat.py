# flake8: noqa

try:
    from griffe import GriffeLoader
    from griffe import ModulesCollection, LinesCollection

    import _griffe.models as dataclasses
    import _griffe.docstrings.models as docstrings
    import _griffe.expressions as expressions

    from griffe import Parser, parse, parse_numpy
    from griffe import AliasResolutionError

    from griffe import load_extensions
except ImportError:
    from griffe.loader import GriffeLoader
    from griffe.collections import ModulesCollection, LinesCollection

    import griffe.dataclasses as dataclasses
    import griffe.docstrings.dataclasses as docstrings
    import griffe.expressions as expressions

    from griffe.docstrings.parsers import Parser, parse
    from griffe.exceptions import AliasResolutionError

    from griffe import load_extensions
