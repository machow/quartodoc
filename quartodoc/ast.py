from __future__ import annotations

import warnings

from enum import Enum
from dataclasses import dataclass
from griffe.docstrings import dataclasses as ds
from griffe import dataclasses as dc
from plum import dispatch
from typing import Union
from pydantic import BaseModel  # for previewing


# Transform and patched-in classes ============================================
# TODO: annotate transform return types. make sure subtypes inherit from correct
# griffe base objects.
# TODO: it seems like transform should happen on the root, not individual elements.


def transform(el):
    """Return a more specific docstring element, or simply return the original one."""

    if isinstance(el, tuple):
        try:
            return tuple_to_data(el)
        except ValueError:
            pass

    # patch a list of docstring sections. note that this has to happen on the
    # list, since we replace single nodes on the tree (the list is the node).
    elif isinstance(el, list) and len(el) and isinstance(el[0], ds.DocstringSection):
        return _DocstringSectionPatched.transform_all(el)

    return el


# Patch DocstringSection ------------------------------------------------------


class DocstringSectionKindPatched(Enum):
    see_also = "see also"
    notes = "notes"
    warnings = "warnings"


class _DocstringSectionPatched(ds.DocstringSection):
    _registry: "dict[Enum, _DocstringSectionPatched]" = {}

    def __init__(self, value: str, title: "str | None" = None):
        super().__init__(title)
        self.value = value

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        if cls.kind.value in cls._registry:
            raise KeyError(f"A section for kind {cls.kind} already exists")

        cls._registry[cls.kind.value] = cls

    @staticmethod
    def split_sections(text) -> list[tuple[str, str]]:
        """Return tuples of (title, body) for all numpydoc style sections in the text.

        Note that this function does not check the value of the section header,
        only that it is a header on one line, with dashed on the next.
        """
        import re

        comp = re.compile(r"^([\S \t]+)\n-+$\n?", re.MULTILINE)

        crnt_match = comp.search(text)
        crnt_pos = 0

        # This loop takes a match, then attempts to look ahead at the next one,
        # in order to find the body of text between multiple sections.
        results = []
        while crnt_match is not None:
            next_pos = crnt_pos + crnt_match.end()
            substr = text[next_pos:]
            next_match = comp.search(substr)

            title = crnt_match.groups()[0]
            body = substr if next_match is None else substr[: next_match.start()]

            results.append((title, body))

            crnt_match, crnt_pos = next_match, next_pos

        return results

    @classmethod
    def transform(cls, el: ds.DocstringSection) -> list[ds.DocstringSection]:
        """Attempt to cast DocstringSection element to more specific section type.

        Note that this is meant to patch cases where the general DocstringSectionText
        class represents a section like See Also, etc..
        """

        if not isinstance(el, ds.DocstringSectionText):
            return [el]

        splits = cls.split_sections(el.value)
        results = []
        for title, body in splits:
            sub_cls = cls._registry.get(title.lower(), ds.DocstringSectionText)

            # note that griffe currently doesn't store the title anywhere,
            # but we add the exact title here, so we can be flexible about the
            # sections we parse (e.g. Note instead of Notes) in the future.
            results.append(sub_cls(body, title))

        return results or [el]

    @classmethod
    def transform_all(cls, el: list[ds.DocstringSection]) -> list[ds.DocstringSection]:
        return sum(map(cls.transform, el), [])


class DocstringSectionSeeAlso(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.see_also


class DocstringSectionNotes(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.notes


class DocstringSectionWarnings(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.warnings


# Patch Example elements ------------------------------------------------------


@dataclass
class ExampleCode:
    value: str


@dataclass
class ExampleText:
    value: str


def tuple_to_data(el: "tuple[ds.DocstringSectionKind, str]"):
    """Re-format funky tuple setup in example section to be a class."""
    assert len(el) == 2

    kind, value = el
    if kind.value == "examples":
        return ExampleCode(value)
    elif kind.value == "text":
        return ExampleText(value)

    raise ValueError(f"Unsupported first element in tuple: {kind}")


# Tree previewer ==============================================================


@dispatch
def fields(el: BaseModel):
    # TODO: this is the only quartodoc specific code.
    # pydantic seems to copy MISSING() when it's a default, so we can't
    # whether a MISSING() is the default MISSING(). Instead, we'll just
    # use isinstance for this particular class
    from .layout import MISSING

    # return fields whose values were not set to the default
    field_defaults = {mf.name: mf.default for mf in el.__fields__.values()}
    return [
        k for k, v in el if field_defaults[k] is not v if not isinstance(v, MISSING)
    ]


@dispatch
def fields(el: dc.Object):
    options = [
        "name",
        "canonical_path",
        "classes",
        "parameters",
        "members",
        "functions",
        "docstring",
    ]
    return [opt for opt in options if hasattr(el, opt)]


@dispatch
def fields(el: dc.ObjectAliasMixin):
    try:
        return fields(el.target)
    except dc.AliasResolutionError:
        warnings.warn(
            f"Could not resolve Alias target `{el.target_path}`."
            " This often occurs because the module was not loaded (e.g. it is a"
            " package outside of your package)."
        )
        return ["name", "target_path"]


@dispatch
def fields(el: dc.Function):
    return ["name", "annotation", "parameters", "docstring"]


@dispatch
def fields(el: dc.Attribute):
    return ["name", "annotation"]


@dispatch
def fields(el: dc.Docstring):
    return ["parser", "parsed"]


@dispatch
def fields(el: ds.DocstringSection):
    return ["kind", "title", "value"]


@dispatch
def fields(el: ds.DocstringParameter):
    return ["annotation", "default", "description", "name", "value"]


@dispatch
def fields(el: ds.DocstringElement):
    return ["annotation", "description"]


@dispatch
def fields(el: ds.DocstringNamedElement):
    return ["name", "annotation", "description"]


@dispatch
def fields(el: dc.Parameter):
    return ["annotation", "kind", "name", "default"]


@dispatch
def fields(el: dict):
    return list(el.keys())


@dispatch
def fields(el: Union[list, dc.Parameters]):
    return list(range(len(el)))


@dispatch
def fields(el: object):
    if isinstance(el, (ExampleCode, ExampleText)):
        from dataclasses import fields as _fields

        return [field.name for field in _fields(el)]

    return None


class Formatter:
    n_spaces = 3
    icon_block = "█─"
    icon_pipe = "├─"
    icon_endpipe = "└─"
    icon_connector = "│ "
    string_truncate_mark = " ..."

    def __init__(self, string_max_length: int = 50, max_depth=999, compact=False):
        self.string_max_length = string_max_length
        self.max_depth = max_depth
        self.compact = compact

    def format(self, call, depth=0, pad=0):
        """Return a Symbolic or Call back as a nice tree, with boxes for nodes."""

        call = transform(call)

        crnt_fields = fields(call)

        if crnt_fields is None:
            str_repr = repr(call)
            if len(str_repr) > self.string_max_length:
                return str_repr[: self.string_max_length] + self.string_truncate_mark

            return str_repr

        call_str = self.icon_block + call.__class__.__name__

        # short-circuit for max depth ----
        if depth >= self.max_depth:
            return call_str + self.string_truncate_mark

        # format arguments ----
        fields_str = []
        for name in crnt_fields:
            val = self.get_field(call, name)

            # either align subfields with the end of the name, or put the node
            # on a newline, so it doesn't have to be so indented.
            if self.compact:
                sub_pad = pad
                linebreak = "\n" if fields(val) else ""
            else:
                sub_pad = len(str(name)) + self.n_spaces
                linebreak = ""

            # do formatting
            formatted_val = self.format(val, depth + 1, pad=sub_pad)
            fields_str.append(f"{name} = {linebreak}{formatted_val}")

        padded = []
        for ii, entry in enumerate(fields_str):
            is_final = ii == len(fields_str) - 1

            chunk = self.fmt_pipe(entry, is_final=is_final, pad=pad)
            padded.append(chunk)

        return "".join([call_str, *padded])

    def get_field(self, obj, k):
        if isinstance(obj, (dict, list, dc.Parameters)):
            return obj[k]

        return getattr(obj, k)

    def fmt_pipe(self, x, is_final=False, pad=0):
        if not is_final:
            connector = self.icon_connector if not is_final else "  "
            prefix = self.icon_pipe
        else:
            connector = "  "
            prefix = self.icon_endpipe

        connector = "\n" + " " * pad + connector
        prefix = "\n" + " " * pad + prefix
        # NOTE: because visiting is depth first, this is essentially prepending
        # the text to the left.
        return prefix + connector.join(x.splitlines())


def preview(
    ast: "dc.Object | ds.Docstring | object",
    max_depth=999,
    compact=False,
    as_string: bool = False,
):
    """Print a friendly representation of a griffe object (e.g. function, docstring)

    Examples
    --------

    >>> from quartodoc import get_object
    >>> obj = get_object("quartodoc", "get_object")

    >>> preview(obj.docstring.parsed)
     ...

    >>> preview(obj)
     ...

    """

    res = Formatter(max_depth=max_depth, compact=compact).format(ast)

    if as_string:
        return res

    print(res)
