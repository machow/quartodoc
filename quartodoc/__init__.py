"""quartodoc is a package for building delightful python API documentation.
"""

# flake8: noqa

from .autosummary import get_function, get_object, Builder
from .renderers import MdRenderer
from .inventory import convert_inventory, create_inventory
from .ast import preview
from .builder.blueprint import blueprint
from .builder.collect import collect
from .layout import Auto

__all__ = (
    "Auto",
    "blueprint",
    "collect",
    "convert_inventory",
    "create_inventory",
    "get_object",
    "preview",
    "Builder",
    "BuilderPkgdown",
    "BuilderSinglePage",
    "MdRenderer",
)
