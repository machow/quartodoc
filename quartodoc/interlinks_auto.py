from __future__ import annotations

import jedi

import ast
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn

from typing_extensions import TypeAlias, Union

ASTItems: TypeAlias = Union[ast.Call, ast.Name, ast.Attribute]


class CallVisitor(ast.NodeVisitor):
    results: list[ASTItems]

    def __init__(self):
        self.results = []

    def visit_Call(self, node):
        if isinstance(node.func, ast.Attribute):
            self.results.append(node)
            return self.visit(node.func.value)
        elif isinstance(node.func, ast.Name):
            self.results.append(node)

    def visit_Name(self, node):
        self.results.append(node)

    def visit_Attribute(self, node):
        self.results.append(node)

    def reset(self):
        self.results = []

        return self

    @classmethod
    def analyze(cls, code) -> list[Interlink]:
        visitor = cls()
        visitor.visit(ast.parse(code))
        script = jedi.Script(code)
        all_links = [node_to_interlink_name(script, call) for call in visitor.results]
        return [link for link in all_links if link is not None]


def narrow_node_start(node: ast.AST) -> tuple[int, int]:
    if isinstance(node, ast.Attribute):
        return (node.value.end_lineno, node.value.end_col_offset)

    return node.lineno, node.col_offset


def node_to_interlink_name(script: jedi.Script, node: ast.AST) -> Interlink | None:
    print(node.func.lineno, node.func.end_col_offset)
    try:
        func = node.func
        name = script.goto(func.lineno, func.end_col_offset)[0]
        full_name = name.full_name
        lineno, col_offset = narrow_node_start(func)
        return Interlink(
            name=full_name,
            lineno=lineno,
            end_lineno=func.end_lineno,
            col_offset=col_offset,
            end_col_offset=func.end_col_offset,
        )
    except IndexError:
        return None


class Code(BaseModel):
    content: str


class Interlink(BaseModel):
    name: str
    lineno: int
    end_lineno: int
    col_offset: int
    end_col_offset: int


app = FastAPI()


@app.post("/analyze")
async def analyze(code: Code):
    res = CallVisitor.analyze(code.content)
    return {"interlinks": res}


if __name__ == "__main__":
    import sys

    port = int(sys.argv[1])
    uvicorn.run(app, port=port, host="0.0.0.0")
