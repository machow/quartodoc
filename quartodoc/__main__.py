import click
import contextlib
import os
import time
import sphobjinv as soi
import yaml
import importlib
from pathlib import Path
from watchdog.observers import Observer
from functools import partial
from watchdog.events import PatternMatchingEventHandler
from quartodoc import Builder, convert_inventory
from pydantic import BaseModel


def get_package_path(package_name):
    """
    Get the path to a package installed in the current environment.
    """
    try:
        lib = importlib.import_module(package_name)
        return lib.__path__[0]
    except ModuleNotFoundError:
        raise ModuleNotFoundError(
            f"Package {package_name} not found.  Please install it in your environment."
        )


class FileInfo(BaseModel):
    size: int
    mtime: float
    name: str = ""


class QuartoDocFileChangeHandler(PatternMatchingEventHandler):
    """
    A handler for file changes.
    """

    # Ignore patterns for the file watcher that are not relevant to the docs
    py_ignore_patterns = [
        "*/__pycache__/*",  # These are the compiled python code files which are automatically generated by Python
        "*/.ipynb_checkpoints/*",  # This directory is created by Jupyter Notebook for auto-saving notebooks
        "*/.vscode/*",  # If you're using Visual Studio Code, it creates this directory to store settings specific to that project.
        "*/.idea/*",  # Similar to .vscode/, but for JetBrains IDEs like PyCharm.
        "*/.git/*",  # i This directory is created by Git. It's not relevant to the docs.
        "*/venv/*",
        "*/env/*",
        "*/.env/*",  # Common names for directories containing a Python virtual environment.
        "*/.pytest_cache/*",  # This directory is created when you run Pytest.
        "*/.eggs/*",
        "*/dist/*",
        "*/build/*",
        "*/*.egg-info/*",  # These are typically created when building Python packages with setuptools.
        "*.pyo",  # These are optimized .pyc files, created when Python is run with the -O flag.
        "*.pyd",  # This is the equivalent of a .pyc file, but for C extensions on Windows.
        "*/.mypy_cache/*",  # This directory is created when you run mypy.
    ]

    def __init__(self, callback):
        super().__init__(
            ignore_patterns=self.py_ignore_patterns, ignore_directories=True
        )
        self.callback = callback
        self.old_file_info = FileInfo(size=-1, mtime=-1, name="")

    def get_file_info(self, path: str) -> FileInfo:
        """
        Get the file size and modification time.
        """
        return FileInfo(
            size=os.stat(path).st_size, mtime=os.stat(path).st_mtime, name=path
        )

    def is_diff(self, old: FileInfo, new: FileInfo) -> bool:
        """
        Check if a file has changed. Prevents duplicate events from being triggered.
        """
        same_nm = old.name == new.name
        diff_sz = old.size != new.size
        diff_tm = new.mtime - old.mtime  # wait 1/4 second before triggering

        if (
            diff_tm < 0.25
        ):  # if consequetive events are less than 1/4th of a second apart, ignore
            return False
        elif same_nm:
            if diff_sz or diff_tm >= 0.25:
                return True
            else:
                return False
        else:
            return True

    def callback_if_diff(self, event):
        """
        Call the callback if the file has changed.
        """
        new_file_info = self.get_file_info(event.src_path)
        if self.is_diff(self.old_file_info, new_file_info):
            self.print_event(event)
            self.callback()
        self.old_file_info = new_file_info

    @classmethod
    def print_event(cls, event):
        print(f"Rebuilding docs.  Detected: {event.event_type} path : {event.src_path}")

    def on_modified(self, event):
        self.callback_if_diff(event)

    def on_created(self, event):
        self.callback_if_diff(event)


def _enable_logs():
    import logging
    import sys

    root = logging.getLogger("quartodoc")
    root.setLevel(logging.INFO)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    handler.setFormatter(formatter)
    root.addHandler(handler)


@contextlib.contextmanager
def chdir(new_dir):
    prev = os.getcwd()
    os.chdir(new_dir)
    try:
        yield new_dir
    finally:
        os.chdir(prev)


@click.group()
def cli():
    pass


@click.command()
@click.option(
    "--config",
    default="_quarto.yml",
    help="Change the path to the configuration file.  The default is `./_quarto.yml`",
)
@click.option(
    "--filter",
    nargs=1,
    default="*",
    help="Specify the filter to select specific files. The default is '*' which selects all files.",
)
@click.option(
    "--dry-run",
    is_flag=True,
    default=False,
    help="If set, prevents new documents from being generated.",
)
@click.option(
    "--watch",
    is_flag=True,
    default=False,
    help="If set, the command will keep running and watch for changes in the package directory.",
)
@click.option("--verbose", is_flag=True, default=False, help="Enable verbose logging.")
def build(config, filter, dry_run, watch, verbose):
    """
    Generate API docs based on the given configuration file  (`./_quarto.yml` by default).
    """
    cfg_path = f"{os.getcwd()}/{config}"
    if not Path(cfg_path).exists():
        raise FileNotFoundError(
            f"Configuration file {cfg_path} not found.  Please create one."
        )
    if verbose:
        _enable_logs()

    builder = Builder.from_quarto_config(config)
    doc_build = partial(builder.build, filter=filter)

    if dry_run:
        pass
    else:
        with chdir(Path(config).parent):
            if watch:
                pkg_path = get_package_path(builder.package)
                print(f"Watching {pkg_path} for changes...")
                observer = Observer()
                observer._event_queue.maxsize = 1  # the default is 0 which is infinite, and there isn't a way to set this in the constructor
                event_handler = QuartoDocFileChangeHandler(callback=doc_build)
                observer.schedule(event_handler, pkg_path, recursive=True)
                observer.schedule(event_handler, cfg_path, recursive=True)
                observer.start()
                try:
                    while True:
                        time.sleep(1)
                except KeyboardInterrupt:
                    pass
                finally:
                    observer.stop()
                    observer.join()
            else:
                doc_build()


@click.command()
@click.argument("config", default="_quarto.yml")
@click.option("--dry-run", is_flag=True, default=False)
def interlinks(config, dry_run):
    cfg = yaml.safe_load(open(config))
    interlinks = cfg.get("interlinks", None)

    cache = cfg.get("cache", "_inv")

    p_root = Path(config).parent

    if interlinks is None:
        print("No interlinks field found in your quarto config. Quitting.")
        return

    for k, v in interlinks["sources"].items():
        # TODO: user shouldn't need to include their own docs in interlinks
        if v["url"] == "/":
            continue

        url = v["url"] + v.get("inv", "objects.inv")
        inv = soi.Inventory(url=url)

        p_dst = p_root / cache / f"{k}_objects.json"
        p_dst.parent.mkdir(exist_ok=True, parents=True)

        convert_inventory(inv, p_dst)


cli.add_command(build)
cli.add_command(interlinks)


if __name__ == "__main__":
    cli()
