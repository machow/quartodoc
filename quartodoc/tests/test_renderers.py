import pytest
import griffe.docstrings.dataclasses as ds
import griffe.expressions as exp

from quartodoc.renderers import MdRenderer
from quartodoc import layout, get_object, blueprint, Auto


@pytest.fixture
def renderer():
    return MdRenderer()


def test_render_param_kwargs(renderer):
    f = get_object("quartodoc.tests.example_signature.no_annotations")
    res = renderer.render(f.parameters)

    assert res == "a, b=1, *args, c, d=2, **kwargs"


def test_render_param_kwargs_annotated():
    renderer = MdRenderer(show_signature_annotations=True)
    f = get_object("quartodoc.tests.example_signature.yes_annotations")

    res = renderer.render(f.parameters)

    assert (
        res
        == "a: int, b: int = 1, *args: list\\[str\\], c: int, d: int, **kwargs: dict\\[str, str\\]"
    )


@pytest.mark.parametrize(
    "src, dst",
    [
        ("example_signature.pos_only", "x, /, a, b=2"),
        ("example_signature.kw_only", "x, *, a, b=2"),
        ("example_signature.early_args", "x, *args, a, b=2, **kwargs"),
        ("example_signature.late_args", "x, a, b=2, *args, **kwargs"),
    ],
)
def test_render_param_kwonly(src, dst, renderer):
    f = get_object("quartodoc.tests", src)

    res = renderer.render(f.parameters)
    assert res == dst


@pytest.mark.parametrize(
    "pair",
    [
        [ds.DocstringSectionParameters, ds.DocstringParameter],
        [ds.DocstringSectionAttributes, ds.DocstringAttribute],
        [ds.DocstringSectionReturns, ds.DocstringReturn],
    ],
)
def test_render_table_description_interlink(renderer, pair):
    interlink = "[](`abc`)"
    cls_section, cls_par = pair
    pars = cls_section([cls_par(name="x", description=interlink)])

    res = renderer.render(pars)
    assert interlink in res


@pytest.mark.parametrize(
    "section, dst",
    [
        (layout.Section(title="abc"), "## abc"),
        (layout.Section(subtitle="abc"), "### abc"),
        (layout.Section(title="abc", desc="zzz"), "## abc\n\nzzz"),
        (layout.Section(subtitle="abc", desc="zzz"), "### abc\n\nzzz"),
    ],
)
def test_render_summarize_section_title(renderer, section, dst):
    res = renderer.summarize(section)

    assert res == dst


def test_render_summarize_section_contents(renderer):
    obj = blueprint(layout.Auto(name="a_func", package="quartodoc.tests.example"))
    section = layout.Section(title="abc", desc="zzz", contents=[obj])
    res = renderer.summarize(section)

    table = (
        "| | |\n| --- | --- |\n"
        "| [a_func](#quartodoc.tests.example.a_func) | A function |"
    )
    assert res == f"## abc\n\nzzz\n\n{table}"


def test_render_doc_attribute(renderer):
    attr = ds.DocstringAttribute(
        name="abc",
        description="xyz",
        annotation=exp.Expression(exp.Name("Optional", full="Optional"), "[", "]"),
        value=1,
    )

    res = renderer.render(attr)

    assert res == ["abc", r"Optional\[\]", "xyz"]


@pytest.mark.parametrize("children", ["embedded", "flat"])
def test_render_doc_module(snapshot, renderer, children):
    bp = blueprint(Auto(name="quartodoc.tests.example", children=children))
    res = renderer.render(bp)

    assert res == snapshot


@pytest.mark.parametrize("children", ["embedded", "flat"])
def test_render_doc_class(snapshot, renderer, children):
    bp = blueprint(Auto(name="quartodoc.tests.example_class.C", children=children))
    res = renderer.render(bp)

    assert res == snapshot


def test_render_doc_class_attributes_section(snapshot, renderer):
    bp = blueprint(Auto(name="quartodoc.tests.example_class.AttributesTable"))
    res = renderer.render(bp)

    assert res == snapshot


@pytest.mark.parametrize("parser", ["google", "numpy", "sphinx"])
def test_render_docstring_styles(snapshot, renderer, parser):
    package = "quartodoc.tests.example_docstring_styles"
    auto = Auto(name=f"f_{parser}", package=package)
    bp = blueprint(auto, parser=parser)

    res = renderer.render(bp)

    assert res == snapshot


def test_render_docstring_numpy_linebreaks(snapshot, renderer):
    package = "quartodoc.tests.example_docstring_styles"
    auto = Auto(name="f_numpy_with_linebreaks", package=package)
    bp = blueprint(auto)

    res = renderer.render(bp)

    assert res == snapshot


def test_render_doc_signature_path(snapshot, renderer):
    package = "quartodoc.tests"
    auto = Auto(name="example.a_func", package=package, signature_path="short")
    bp = blueprint(auto)

    res = renderer.render(bp)

    assert res == snapshot


def test_render_source(renderer):
    obj = get_object("quartodoc.tests.example.a_func")
    res = renderer.render_source(obj)

    code = '''\
def a_func():
    """A function"""\
'''

    res == f"<code><pre>{code}</pre></code>"


def test_render_doc_show_source(renderer, snapshot):
    package = "quartodoc.tests.example_docstring_styles"
    auto = Auto(name="f_quarto_block_in_docstring", package=package, show_source=True)
    bp = blueprint(auto)

    res = renderer.render(bp)

    assert res == snapshot
