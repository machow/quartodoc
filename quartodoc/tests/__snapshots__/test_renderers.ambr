# serializer version: 1
# name: test_render_doc_class[embedded]
  '''
  # quartodoc.tests.example_class.C { #quartodoc.tests.example_class.C }
  
  `tests.example_class.C(self, x, y)`
  
  The short summary.
  
  The extended summary,
  which may be multiple lines.
  
  ## Parameters
  
  | Name   | Type   | Description          | Default    |
  |--------|--------|----------------------|------------|
  | `x`    | str    | Uses signature type. | _required_ |
  | `y`    | int    | Uses manual type.    | _required_ |
  
  ## Attributes
  
  | Name | Description |
  | --- | --- |
  | [SOME_ATTRIBUTE](#quartodoc.tests.example_class.C.SOME_ATTRIBUTE) | An attribute |
  | [some_property](#quartodoc.tests.example_class.C.some_property) | A property |
  | [z](#quartodoc.tests.example_class.C.z) | A documented init attribute |
  
  ## Classes
  
  | Name | Description |
  | --- | --- |
  | [D](#quartodoc.tests.example_class.C.D) | A nested class |
  
  ### D { #quartodoc.tests.example_class.C.D }
  
  `tests.example_class.C.D()`
  
  A nested class
  
  ## Methods
  
  | Name | Description |
  | --- | --- |
  | [some_method](#quartodoc.tests.example_class.C.some_method) | A method |
  
  ### some_method { #quartodoc.tests.example_class.C.some_method }
  
  `tests.example_class.C.some_method(self)`
  
  A method
  '''
# ---
# name: test_render_doc_class[flat]
  '''
  # quartodoc.tests.example_class.C { #quartodoc.tests.example_class.C }
  
  `tests.example_class.C(self, x, y)`
  
  The short summary.
  
  The extended summary,
  which may be multiple lines.
  
  ## Parameters
  
  | Name   | Type   | Description          | Default    |
  |--------|--------|----------------------|------------|
  | `x`    | str    | Uses signature type. | _required_ |
  | `y`    | int    | Uses manual type.    | _required_ |
  
  ## Attributes
  
  | Name | Description |
  | --- | --- |
  | [SOME_ATTRIBUTE](#quartodoc.tests.example_class.C.SOME_ATTRIBUTE) | An attribute |
  | [some_property](#quartodoc.tests.example_class.C.some_property) | A property |
  | [z](#quartodoc.tests.example_class.C.z) | A documented init attribute |
  
  ## Classes
  
  | Name | Description |
  | --- | --- |
  | [D](#quartodoc.tests.example_class.C.D) | A nested class |
  
  ## D { #quartodoc.tests.example_class.C.D }
  
  `tests.example_class.C.D()`
  
  A nested class
  
  ## Methods
  
  | Name | Description |
  | --- | --- |
  | [some_method](#quartodoc.tests.example_class.C.some_method) | A method |
  
  ## some_method { #quartodoc.tests.example_class.C.some_method }
  
  `tests.example_class.C.some_method(self)`
  
  A method
  '''
# ---
# name: test_render_doc_class_attributes_section
  '''
  # quartodoc.tests.example_class.AttributesTable { #quartodoc.tests.example_class.AttributesTable }
  
  `tests.example_class.AttributesTable(self)`
  
  The short summary.
  
  ## Attributes
  
  | Name   | Type   | Description         |
  |--------|--------|---------------------|
  | x      | str    | Uses signature type |
  | y      | int    | Uses manual type    |
  | z      | float  | Defined in init     |
  '''
# ---
# name: test_render_doc_module[embedded]
  '''
  # quartodoc.tests.example { #quartodoc.tests.example }
  
  `tests.example`
  
  A module
  
  ## Attributes
  
  | Name | Description |
  | --- | --- |
  | [a_attr](#quartodoc.tests.example.a_attr) | An attribute |
  
  ## Classes
  
  | Name | Description |
  | --- | --- |
  | [AClass](#quartodoc.tests.example.AClass) | A class |
  
  ### AClass { #quartodoc.tests.example.AClass }
  
  `tests.example.AClass()`
  
  A class
  
  #### Attributes
  
  | Name | Description |
  | --- | --- |
  | [a_attr](#quartodoc.tests.example.AClass.a_attr) | A class attribute |
  
  #### Methods
  
  | Name | Description |
  | --- | --- |
  | [a_method](#quartodoc.tests.example.AClass.a_method) | A method |
  
  ##### a_method { #quartodoc.tests.example.AClass.a_method }
  
  `tests.example.AClass.a_method()`
  
  A method
  
  ## Functions
  
  | Name | Description |
  | --- | --- |
  | [a_func](#quartodoc.tests.example.a_func) | A function |
  
  ### a_func { #quartodoc.tests.example.a_func }
  
  `tests.example.a_func()`
  
  A function
  '''
# ---
# name: test_render_doc_module[flat]
  '''
  # quartodoc.tests.example { #quartodoc.tests.example }
  
  `tests.example`
  
  A module
  
  ## Attributes
  
  | Name | Description |
  | --- | --- |
  | [a_attr](#quartodoc.tests.example.a_attr) | An attribute |
  
  ## Classes
  
  | Name | Description |
  | --- | --- |
  | [AClass](#quartodoc.tests.example.AClass) | A class |
  
  ## AClass { #quartodoc.tests.example.AClass }
  
  `tests.example.AClass()`
  
  A class
  
  ### Attributes
  
  | Name | Description |
  | --- | --- |
  | [a_attr](#quartodoc.tests.example.AClass.a_attr) | A class attribute |
  
  ### Methods
  
  | Name | Description |
  | --- | --- |
  | [a_method](#quartodoc.tests.example.AClass.a_method) | A method |
  
  #### a_method { #quartodoc.tests.example.AClass.a_method }
  
  `tests.example.AClass.a_method()`
  
  A method
  
  ## Functions
  
  | Name | Description |
  | --- | --- |
  | [a_func](#quartodoc.tests.example.a_func) | A function |
  
  ## a_func { #quartodoc.tests.example.a_func }
  
  `tests.example.a_func()`
  
  A function
  '''
# ---
# name: test_render_doc_show_source
  '''
  # f_quarto_block_in_docstring { #quartodoc.tests.example_docstring_styles.f_quarto_block_in_docstring }
  
  `tests.example_docstring_styles.f_quarto_block_in_docstring(a, b)`
  
  A quarto style docstring.
  
  ```{python}
  1 < 2
  ```
  
  <details>
  <summary>Show source</summary>
  <pre><code>def f_quarto_block_in_docstring(a, b: str):
      """A quarto style docstring.
  
      &#96;&#96;&#96;{python}
      1 &lt; 2
      &#96;&#96;&#96;
      """</code></pre>
  </details>
  '''
# ---
# name: test_render_doc_signature_path
  '''
  # example.a_func { #quartodoc.tests.example.a_func }
  
  `a_func()`
  
  A function
  '''
# ---
# name: test_render_docstring_numpy_linebreaks
  '''
  # f_numpy_with_linebreaks { #quartodoc.tests.example_docstring_styles.f_numpy_with_linebreaks }
  
  `tests.example_docstring_styles.f_numpy_with_linebreaks(a, b)`
  
  A numpy style docstring.
  
  ## Parameters
  
  | Name   | Type   | Description      | Default    |
  |--------|--------|------------------|------------|
  | `a`    |        | The a parameter. | _required_ |
  | `b`    | str    | The b parameter. | _required_ |
  '''
# ---
# name: test_render_docstring_styles[google]
  '''
  # f_google { #quartodoc.tests.example_docstring_styles.f_google }
  
  `tests.example_docstring_styles.f_google(a, b)`
  
  A google style docstring.
  
  ## Parameters
  
  | Name   | Type   | Description      | Default    |
  |--------|--------|------------------|------------|
  | `a`    | int    | The a parameter. | _required_ |
  | `b`    | str    | The b parameter. | _required_ |
  '''
# ---
# name: test_render_docstring_styles[numpy]
  '''
  # f_numpy { #quartodoc.tests.example_docstring_styles.f_numpy }
  
  `tests.example_docstring_styles.f_numpy(a, b)`
  
  A numpy style docstring.
  
  ## Parameters
  
  | Name   | Type   | Description      | Default    |
  |--------|--------|------------------|------------|
  | `a`    |        | The a parameter. | _required_ |
  | `b`    | str    | The b parameter. | _required_ |
  '''
# ---
# name: test_render_docstring_styles[sphinx]
  '''
  # f_sphinx { #quartodoc.tests.example_docstring_styles.f_sphinx }
  
  `tests.example_docstring_styles.f_sphinx(a, b)`
  
  A sphinx style docstring.
  
  ## Parameters
  
  | Name   | Type   | Description      | Default    |
  |--------|--------|------------------|------------|
  | `a`    | int    | The a parameter. | _required_ |
  | `b`    | str    | The b parameter. | _required_ |
  '''
# ---
